### PR TITLE
[Feature] MongoDB `bulk_write` operations for `/api/kytos/flow_manager/v2/flows/`

### LINKED ISSUE #80: MongoDB `bulk_write` operations for `/api/kytos/flow_manager/v2/flows/`
There's opportunity to leverage `bulk_write` on MongoDB for `/api/kytos/flow_manager/v2/flows/` when adding/updating or deleting flows. This will be delivered separately from issue #75 to facilitate reviewing.

### PR DESCRIPTION
**Pull request by viniarck**
Fixes #80 

### Description of the change

- Refactored `flows` upsert and delete operations to use `bulk_write` instead for higher performance based on the expected workload
- Endpoint /flow_manager/v2/flows/ writes first to the database now to optimize consistency for bulk operations

This table summarize the ratio of performance gains comparing this branch `feature/bulk_ops` with `feature/mongo`, the number of DB operations are no longer proportional of the number of flows in the body but constant to the number of the request. For upserts (update or insert) with 3 flows, the gains are significant, 100x faster comparing the 95 percentile. For deletions, it's 10x faster comparing the 95 percentile. For a single flow it's relatively the same:

2022-05-26-124839_1539x323_scrot

## Stress test

To measure the performance, it was stress tested with 50 req / sec during 60 secs with the following payloads, Elastic APM was also actively used to monitor the DB operators performance (but to minimize the information I'm only posting the primary information here):

### POST 50 req/sec with 1 flows, branch `feature/bulk_ops`

```
❯ jq -ncM '{method: "POST", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": true, "flows": [ { "priority": 10, "match": { "in_port"
 : 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/json"]}}' | vegeta attack -format=json -rate 50/1s -
duration=60s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         3000, 50.02, 50.01
Duration      [total, attack, wait]             59.99s, 59.981s, 9.965ms
Latencies     [min, mean, 50, 90, 95, 99, max]  5.25ms, 21.683ms, 13.116ms, 47.324ms, 64.941ms, 105.676ms, 256.573ms
Bytes In      [total, mean]                     111000, 37.00
Bytes Out     [total, mean]                     366000, 122.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:3000
Error Set:
```

### POST 50 req/sec with 1 flow, branch `feature/mongo`

```
❯ jq -ncM '{method: "POST", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": true, "flows": [ { "priority": 10, "match": { "in_port"
 : 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/json"]}}' | vegeta attack -format=json -rate 50/1s -
duration=60s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         3000, 50.02, 50.01
Duration      [total, attack, wait]             59.99s, 59.98s, 9.965ms
Latencies     [min, mean, 50, 90, 95, 99, max]  5.256ms, 19.966ms, 12.204ms, 45.041ms, 58.235ms, 88.72ms, 150.608ms
Bytes In      [total, mean]                     111000, 37.00
Bytes Out     [total, mean]                     366000, 122.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:3000
Error Set:
```

### POST 50 req/sec with 3 flows, branch `feature/bulk_ops`

```
❯ jq -ncM '{method: "POST", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": true, "flows": [ { "priority": 10, "match": { "in_port"
 : 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 101 }, "actions": [ { "action_type": "output", "por
t": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 102 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/js
on"]}}' | vegeta attack -format=json -rate 50/1s -duration=60s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         3000, 50.02, 50.00
Duration      [total, attack, wait]             59.998s, 59.979s, 18.393ms
Latencies     [min, mean, 50, 90, 95, 99, max]  6.819ms, 27.887ms, 18.707ms, 52.68ms, 75.7ms, 140.392ms, 220.288ms
Bytes In      [total, mean]                     111000, 37.00
Bytes Out     [total, mean]                     954000, 318.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:3000
Error Set:
```

### POST 50 req/sec with 3 flows, branch `feature/mongo`

```
❯ jq -ncM '{method: "POST", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": true, "flows": [ { "priority": 10, "match": { "in_port"
 : 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 101 }, "actions": [ { "action_type": "output", "por
t": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 102 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/js
on"]}}' | vegeta attack -format=json -rate 50/1s -duration=60s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         3000, 50.02, 46.14
Duration      [total, attack, wait]             1m5s, 59.98s, 5.04s
Latencies     [min, mean, 50, 90, 95, 99, max]  15.647ms, 3.779s, 3.92s, 8.09s, 8.639s, 9.485s, 10.683s
Bytes In      [total, mean]                     111000, 37.00
Bytes Out     [total, mean]                     954000, 318.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:3000
Error Set:
```

### DELETE 50 req/sec with 3 flows, branch `feature/bulk_ops`

```
❯ jq -ncM '{method: "DELETE", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": true, "flows": [ { "priority": 10, "match": { "in_por
t" : 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 101 }, "actions": [ { "action_type": "output", "p
ort": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 102 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/
json"]}}' | vegeta attack -format=json -rate 50/1s -duration=60s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         3000, 50.02, 50.01
Duration      [total, attack, wait]             59.994s, 59.98s, 13.311ms
Latencies     [min, mean, 50, 90, 95, 99, max]  4.53ms, 55.919ms, 14.976ms, 84.775ms, 214.355ms, 929.206ms, 1.414s
Bytes In      [total, mean]                     111000, 37.00
Bytes Out     [total, mean]                     954000, 318.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:3000
Error Set:
```

### DELETE 50 req/sec with 3 flows, branch `feature/mongo`

```
❯ jq -ncM '{method: "DELETE", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": true, "flows": [ { "priority": 10, "match": { "in_por
t" : 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 101 }, "actions": [ { "action_type": "output", "p
ort": 1 } ] }, { "priority": 10, "match": { "in_port": 1, "dl_vlan": 102 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/
json"]}}' | vegeta attack -format=json -rate 50/1s -duration=60s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         3000, 50.02, 50.00
Duration      [total, attack, wait]             59.996s, 59.98s, 16.651ms
Latencies     [min, mean, 50, 90, 95, 99, max]  8.645ms, 413.429ms, 51.541ms, 1.36s, 2.169s, 3.661s, 6.399s
Bytes In      [total, mean]                     111000, 37.00
Bytes Out     [total, mean]                     954000, 318.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:3000
Error Set:
```

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -384,44 +386,33 @@ def check_alien_flows(self, switch):
...
+        flow_ids_to_delete = set()
+        cookies = list({int(value.get("cookie", 0)) for value in flow_dicts})
+        for dpid, stored_flows in self.flow_controller.get_flows_by_cookies(
+            list(switches.keys()), cookies
+        ).items():
+            for flow_dict in flow_dicts:
+                for stored_flow in stored_flows:
+                    if stored_flow["flow_id"] in flow_ids_to_delete:
+                        continue
+                    if match_flow(
```
Comment (viniarck) [inline main.py:406]: In the future, we could also offload this non strict match to the DB, the performance is great now, so probably not worth mapping this enhancement or trying to do this now. Having a full compliant OpenFlow non strict match offloaded would be challenging too (especially with masks), but depending in the future if we reduce the scope and support the major types of matches on certain fields that should be doable with indexes and some query expressions that MongoDB provides.

**Review by italovalcy — APPROVED** [review]
It looks great to me! Amazing to see the performance improvements of doing the bulk update over a minor loss when doing individual cases. Just out of curiosity: why did you choose  3 flows per request for evaluating the bulk update? I mean, mef_eline usually submits two flows/request, so I was expecting to see that on the tests. No need for running new tests!

**Comment by viniarck** [comment]
> It looks great to me! Amazing to see the performance improvements of doing the bulk update over a minor loss when doing individual cases. Just out of curiosity: why did you choose 3 flows per request for evaluating the bulk update? I mean, mef_eline usually submits two flows/request, so I was expecting to see that on the tests. No need for running new tests!

Yep, I'm also glad that MongoDB doesn't require additional locks in the application when using this `bulk_write`. 

@italovalcy It turns out that I chose three flows in the payload one extra flow in the body than mef_eline will send in practice, I've also tested with up to 5 flows but didn't publish all of the results when I was exploring the solution.

**Comment by viniarck** [comment]
Thanks for reviewing, @italovalcy

**Review by ajoaoff — APPROVED** [review]
LGTM

**Pull request merged**